kernel/msg/README.md

```diff
@@ -0,0 +1,211 @@
+				oid_t oid;
+			} create;
+
+			/* LOOKUP */
+			struct {
+				oid_t fil;
+				oid_t dev;
+			} lookup;
+
+			unsigned char raw[64];
+		};
+
+		int err;
+		size_t size;
+		void *data;
+	} o;
+
+} msg_t;
+```
+
+Common fields for all message types are:
+
+- `type` - message type,
+- `pid` - sender process ID,
+- `priority` - priority of sender thread,
+- `oid` - object identifier of the target object (e.g. file or device).
+
+Additionally, there are two fields used to store input and output data specific to the message type:
+
+- `i` - input data,
+- `o` - output data.
+
+Message error codes are stored in the `o.err` field common for all message types.
+
+### Message types
+
+#### `mtOpen`
+
+Open a file or device specified by `msg.oid`.
+
+Input parameters:
+
+- `i.openclose.flags` - flags with which file is being opened.
+
+#### `mtClose`
+
+Close a file or device specified by `msg.oid`.
+
+#### `mtRead`
+
+Read from a file or device specified by `msg.oid`.
+
+Parameters set by caller:
+
+- `i.io.offs` - offset in the file,
+- `i.io.len` - number of bytes to read,
+- `i.io.mode` - flags with which file has been opened.
+- `o.data` - buffer for data,
+- `o.size` - length of the `o.data` buffer.
+
+The operation should block the client until all requested data becomes available. Number of read bytes or error is
+returned via `o.err`.
+
+#### `mtWrite`
+
+Write to a file or device specified by `msg.oid`.
+
+Input parameters:
+
+- `i.io.offs` - offset in the file,
+- `i.io.len` - number of bytes to write,
+- `i.io.mode` - flags with which file has been opened.
+- `i.data` - buffer with data,
+- `i.size` - length of the `i.data` buffer.
+
+The operation should block the client until all requested data is written to the device. Number of written bytes or
+error is returned via `o.err`.
+
+#### `mtTruncate`
+
+Truncate a file.
+
+#### `mtDevCtl`
+
+Custom message type for device control.
+
+This message type allows defining an entirely custom structure for communication with a server. This structure should
+be serialized/deserialized to/from message i.raw/o.raw fields. Additional data can be passed in i.data and o.data
+fields.
+
+#### `mtCreate`
+
+Create new object.
+
+#### `mtDestroy`
+
+Destroy object.
+
+#### `mtSetAttr`
+
+Set object attributes.
+
+#### `mtGetAttr`
+
+Get object attribute specified by `i.attr.type`.
+
+#### `mtGetAttrAll`
+
+Get all object attributes.
+
+#### `mtLookup`
+
+Lookup for object identifier associated with a name.
+
+#### `mtLink`
+
+Create a hard link.
+
+#### `mtUnlink`
+
+Remove a hard link.
+
+#### `mtReaddir`
+
+Read directory entries.
+
+## Message passing
+
+Messages are sent using the `msgSend(port_t port, msg_t *msg)` syscall. The `port` argument is the port number of the
+target queue, and `msg` is a pointer to the message structure. The caller is responsible for filling the message
+structure with appropriate data. Only the `mtDevCtl` message type may use `msg.i.raw` field directly. For all other
```

Comment:
Maybe "msg.i.raw and msg.o.raw must not be used directly when using standard message type, with exception of mtDevCtl"?
```suggestion
structure with appropriate data. Only the `mtDevCtl` message type may use `msg.i.raw` and `msg.i.raw` fields directly. For all other
```